scripts/refine-oauth-token-timeouts.sh

```diff
@@ -0,0 +1,209 @@
+TOKEN_INACTIVITY_TIMEOUT="${1:-300}"
+
+# Verify provided TOKEN_INACTIVITY_TIMEOUT value is a number greater than 300
+if   [[ ! "${TOKEN_INACTIVITY_TIMEOUT}" =~ ^[0-9]{3,}$ ]]
+then
+  echo "Please specify inactivity timeout in seconds without the 's' suffix!"
+  exit 1
+elif [[ ! "${TOKEN_INACTIVITY_TIMEOUT}" -ge "300"      ]]
+then
+  echo "Minimum allowed token inactivity timeout value is 300 (seconds)."
+  echo "Please specify a number equal or greater than that!"
+  exit 1
+fi
+
+# Value of token (session) expiration timeout in seconds of type integer number
+# without the 's' suffix. Optional argument. Defaults to value of
+# TOKEN_INACTIVITY_TIMEOUT multiplied by 10, if unset.
+#
+# TODO / RFE: Explore the possibility to derive this from the actual value of
+#             RH-SSO's 'SSO Session Max' setting
+TOKEN_EXPIRATION_TIMEOUT="${2:-$(( $TOKEN_INACTIVITY_TIMEOUT * 10 ))}"
+
+# Verify the specified TOKEN_EXPIRATION_TIMEOUT value is a number
+if   [[ ! "${TOKEN_EXPIRATION_TIMEOUT}" =~ ^[0-9]{3,}$ ]]
+then
+  echo "Please specify expiration timeout in seconds without the 's' suffix!"
+  exit 1
+fi
+
+# If OpenShift version is v4 and greater set the token expiration timeout for
+# internal OAuth server.
+#
+# Note: Skip this step for OpenShift server of v3, since in that case the
+# timeouts need to be configured in the 'oauthConfig' section of the master
+# configuration file!
+readonly ocServerVersion=$(
+  oc version -o yaml | grep 'openshiftVersion' | grep -oP '((\d+\.){2}\d+)'
+)
+
+echo
+echo "Updating OAuth access token settings:"
+echo
+
+if [[ "${ocServerVersion}" =~ ^4\..+$ ]]
+then
+  show_label "For OpenShift '$ocServerVersion' cluster. Setting token:"
+  echo "* Expiration timeout to: '${TOKEN_EXPIRATION_TIMEOUT}' seconds."
+  oc_patch_token_property "OAuth" "cluster"      \
+    "/spec/tokenConfig/accessTokenMaxAgeSeconds" \
+    "${TOKEN_EXPIRATION_TIMEOUT}"
+  echo "* Inactivity timeout to: '${TOKEN_INACTIVITY_TIMEOUT}' seconds."
+  # Per .spec.tokenConfig config.openshift.io/v1 OAuth API object the datatype
+  # for 'accessTokenInactivityTimeout' property is string, thus append 's' to
+  # the specified value and enclose it with double quotes
+  oc_patch_token_property "OAuth" "cluster"          \
+    "/spec/tokenConfig/accessTokenInactivityTimeout" \
+    "${TOKEN_INACTIVITY_TIMEOUT}s"
+  echo
+else
+  echo "Failed to determine the version of the OpenShift server!"
+  echo "Please login and retry."
+  exit
+fi
+
+# Since OAuth client settings if configured override the same settings possibly
+# present in the internal OAuth server, also configure token expiration and
+# inactivity timeouts for all relevant OAuth clients (console, the default OAuth
+# clients, and those provisioned by a RHOAM instance)
+declare -ra OAUTH_CLIENTS=(
+  "console"
+  "openshift-browser-client"
+  "openshift-challenging-client"
+  "redhat-rhoam-3scale"
+  "redhat-rhoam-rhsso"
+  "redhat-rhoam-rhssouser"
+)
```

Comment:
Could this be possibly parametrized?
Let's say we would want to configure the session and token expiration only for SSO IDP users (not kubeadmin and others)...

Reply:
@psturc Quick clarification regarding this one yet - do you want all of them to be parametrizable? (no default clients to be updated)

Or should some of them by updated by default, but the final list would be created as a concatenation of the default ones plus those specified on the command line? If the latter, can you possibly suggest which of them should be the default ones?

Reply:
These are good questions...

So I suppose this script will be used for both development and testing purposes.

From developer's POV I can imagine that having the session and token expiration the longest would probably make sense, i.e. if I run the SSO IDP script without any extra params, I would expect that the token and session expiration doesn't change at all. (i.e. 24 hours expiration for kubeadmin?)

But from testing POV (where we want to mimic the customer's environment as close as possible), I think that the typical configuration would include configuring the session and token expiration for RHOAM services, i.e. SSO, User-SSO and 3scale.

Else if I specify just the env vars for token and session timeouts, I would expect that all clients would be updated (by default).

So, if it is possible, I would see the final functionality like this:

```bash
# Don't configure token inactivity timeout nor session expiration timeout at all
./scripts/setup-sso-idp.sh

# Set token inactivity and session expiration timeouts for all clients
TOKEN_INACTIVITY_TIMEOUT=300 SESSION_EXPIRATION_TIMEOUT=3600 ./scripts/setup-sso-idp.sh

# Set token inactivity and session expiration timeouts only for 3scale, RHSSO and User SSO
TOKEN_INACTIVITY_TIMEOUT=300 SESSION_EXPIRATION_TIMEOUT=3600 OAUTH_CLIENTS=redhat-rhoam-3scale,redhat-rhoam-rhsso,redhat-rhoam-rhssouser ./scripts/setup-sso-idp.sh
```

Would it make sense?

Reply:
> These are good questions...
> 
> So I suppose this script will be used for both development and testing purposes.

Right.

> 
> From developer's POV I can imagine that having the session and token expiration the longest would probably make sense, i.e. if I run the SSO IDP script without any extra params, I would expect that the token and session expiration doesn't change at all. (i.e. 24 hours expiration for kubeadmin?)

Sounds reasonable. Agree that in development environment, where the cluster management is utilized all the time practically, it would make sense to have a way to keep default OpenShift values (e.g. expiration timeout set to one day, and inactivity timeout not set at all), so the re-logins doesn't need to be issued that often.

> 
> But from testing POV (where we want to mimic the customer's environment as close as possible), I think that the typical configuration would include configuring the session and token expiration for RHOAM services, i.e. SSO, User-SSO and 3scale.

Right, ok.

> 
> Else if I specify just the env vars for token and session timeouts, I would expect that all clients would be updated (by default).

Ok.

> 
> So, if it is possible, I would see the final functionality like this:
> 
> ```shell
> # Don't configure token inactivity timeout nor session expiration timeout at all
> ./scripts/setup-sso-idp.sh
> 
> # Set token inactivity and session expiration timeouts for all clients
> TOKEN_INACTIVITY_TIMEOUT=300 SESSION_EXPIRATION_TIMEOUT=3600 ./scripts/setup-sso-idp.sh
> 
> # Set token inactivity and session expiration timeouts only for 3scale, RHSSO and User SSO
> TOKEN_INACTIVITY_TIMEOUT=300 SESSION_EXPIRATION_TIMEOUT=3600 OAUTH_CLIENTS=redhat-rhoam-3scale,redhat-rhoam-rhsso,redhat-rhoam-rhssouser ./scripts/setup-sso-idp.sh
> ```
> 
> Would it make sense?

Yes, sounds reasonable to me. Thanks for the clarification / sketch about different ways / environments the script could be used. Will mirror above expectations into the code yet.

Reply:
@psturc Above behaviour implemented in the last commit. PTAL once got a chance. Thanks